### PR TITLE
feat: Dealer Interface, 3D playback, equity calculator, Docker Compose

### PR DESCRIPTION
**Pull request by achung123**
### Dealer Interface 

image

Mobile-first dealer UI for live poker sessions with camera-based card capture.

**Components:**
- **GameCreateForm** — Create game sessions with player name + chip inputs
- **HandDashboard** — Hand counter with Submit / New Hand navigation
- **PlayerGrid** — Tappable player tiles showing assigned hole cards
- **CameraCapture** — Camera capture flow with iOS Safari compatibility (user-gesture trigger, no `capture` attr, explicit positioning)
- **DetectionReview** — Review detected cards with confidence scores
- **CardPicker** — 52-card grid for manual correction of misdetected cards

**State Management:**
- `dealerState.js` — useReducer with 5 actions (SET_GAME, SET_PLAYER_CARDS, SET_COMMUNITY_CARDS, RESET_HAND, SET_STEP)
- `handPayload.js` — Payload assembly + duplicate card validation

**API Client:**
- `uploadImage()` — POST multipart image upload
- `getDetectionResults()` — GET detection results by upload ID

**Bug Fixes:**
- iOS Safari: replaced programmatic `.click()` in useEffect with explicit "Open Camera" button (user gesture requirement)
- iOS Safari: removed `capture="environment"` attribute (ignored/broken on iOS)
- iOS Safari: replaced CSS `inset: 0` shorthand with explicit `top/right/bottom/left`
- Fixed `upload.id` → `upload.upload_id` to match backend response field name
- Added `allowedHosts: true` to Vite config for tunnel/remote dev access

**Tests:** 15 frontend unit tests (reducer + payload validation)

---

### YOLO Card Detection — GPU Fine-Tuning & Docker Integration — NEW

image

image

Improved detection accuracy from ~25% to 90%+ confidence on close-up poker hand photos.

**Root Cause (Scale Mismatch):**
Training data contained cards at ~7% of frame (46px in a 640px image); user close-up photos have cards at ~42% of frame. The pretrained model generalized poorly to hands held close to the camera.

**Solution — Fine-tuned Close-up Model (`models/best_closeup.pt`):**
- Generated 4,000 close-up crops (15–50% frame fill) + 1,500 original images from `data/cards/train/`
- Trained 50 epochs on GPU (RTX 3050 Ti) using `scripts/train_card_detector.py`
- **mAP50: 0.950** — AH confidence: 0.249 → 0.905, 2H confidence: 0.752 → 0.881

**YoloCardDetector Updates (`src/app/services/card_detector.py`):**
- Multi-scale inference at imgsz = [480, 640]; best confidence per card value wins
- Confidence threshold lowered to 0.20 (model is well-calibrated after fine-tuning)
- Weight loading priority: `models/best_closeup.pt` → `models/best.pt` → `MockCardDetector`

**Training Script (`scripts/train_card_detector.py`):**
- Reproducible CLI: `--epochs`, `--batch`, `--device`, `--base-model`, `--out-model`, `--closeup-count`, and more
- Freezes first 10 backbone layers for transfer learning efficiency
- Exports TensorBoard logs from `results.csv` at end of run

**TensorBoard:**
- Training metrics visualised at `localhost:6006` via `tensorboard --logdir runs/`

**GPU Support:**
- Replaced `torch 2.2.2+cpu` with `torch 2.6.0+cu124` (CUDA 12.4)

**Docker Fixes:**
- Added `libgl1` + `libglib2.0-0` to Dockerfile (OpenCV runtime dependency)
- Installed `ultralytics` into `.venv` via `uv pip install --python .venv/bin/python`
- Mounted `./models:/app/models` volume in `docker-compose.yml`

---

### 3D Playback View

image

- Three.js table with community cards, hole cards, chip stacks, seat labels, session scrubber, and street scrubber

### Equity Calculator
- Real-time win-probability badges per seat; exhaustive eval for ≤2 unknown cards, Monte Carlo (5,000 iters) otherwise
- Fixed `T`/`10` rank alias bug

### Data View CRUD
- New game, add hand, edit hand, CSV import, and "Load in Visualiser" button

### Docker Compose
- `backend` + `frontend` services; entrypoint runs Alembic migrations then optional seed on startup (`SEED_DATA=1`)

### Seed Script
- `scripts/seed_demo_game.py` creates 3 randomised game sessions; wipes existing data first

### Bug Fixes
- Blank/white screen on startup — `html/body/#app` height chain + canvas dimension guard
- Equity badges not updating between streets — missing `T` entry in `RANK_VAL`

### Infrastructure
- `docker-compose.yml`, `docker-entrypoint.sh`, `frontend/Dockerfile`
- README rewritten with full setup, Docker/Compose, and API reference

## Testing

```bash
# Backend (749 tests)
uv run pytest test/

# Frontend (15 tests)
cd frontend && npx vitest run

# Docker
docker compose up --build

# Fine-tune model (optional, requires GPU)
uv run python scripts/train_card_detector.py --device cuda --epochs 50
```

### CODE REVIEW

**Comment by achung123** [comment]
This is captued in #21, will add a new PR converting to TS + React and merge that.
